### ISSUE TITLE
Control Panel shortcut on task bar? 

### ISSUE THREAD

**Issue by MariuzM** [Disassembler0/Win10-Initial-Setup-Script#147]
:) sorry if i knew how to create it would post it :)

**Comment by Disassembler0** [comment]
**tl;dr:** I hate you. :)

Now for the bedtime story...

First and foremost, I wouldn't accept such tweak into the library. This isn't a setting offered by the basic system, this is already an extra customization which creates something new. I try to generally avoid such and recently refused one [PR which had similar idea](https://github.com/Disassembler0/Win10-Initial-Setup-Script/pull/138). That being said, I wanted to give it a try, just to see if (and how) it can be done.

I spent considerable amount of time researching pinning and unpinning mechanisms already in the past and covered my research in issue #8. Generally, Microsoft tries to make this really hard, so it won't be abused by software vendors and users wouldn't feel that they are not in control. Ironic, I know. Nevertheless, there is a way how it can be done. However your request to pin something what's not a program, but a COM class, makes this even harder.

**Generally speaking:**  
The pinning has been deliberately broken long time ago by KB3093266. Currently, there is pretty neat workaround which looks like this:

```powershell
# Bring pinning / unpinning handler into '*' class scope
$pinHandler = Get-ItemPropertyValue -Path "HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\CommandStore\shell\Windows.taskbarpin" -Name "ExplorerCommandHandler"
New-Item -Path "HKCU:Software\Classes\*\shell\pin" -Force | Out-Null
Set-ItemProperty -LiteralPath "HKCU:Software\Classes\*\shell\pin" -Name "ExplorerCommandHandler" -Type String -Value $pinHandler

# Pin whatever you want to pin
# ....InvokeVerb("pin")

# Remove the handler
Remove-Item -LiteralPath "HKCU:Software\Classes\*\shell\pin" -Recurse
```

This allows you to reintroduce the pinning handler in 'all items' class scope and then invoke it on whatever item you want to pin.

**Approach 1** - Find the namespace of Control Panel and pin it directly.  
The easiest way would be to get the namespace of control panel as a directory item via its CLSID and invoke the method on it.

```powershell
# Get "Control Panel" namespace
$cplCLSID = "::{26EE0668-A00A-44D7-9371-BEB064C98683}"
$cplNS = (New-Object -Com "Shell.Application").NameSpace("shell:$cplCLSID")

# Bring pinning / unpinning handler into '*' class scope
$pinHandler = Get-ItemPropertyValue -Path "HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\CommandStore\shell\Windows.taskbarpin" -Name "ExplorerCommandHandler"
New-Item -Path "HKCU:Software\Classes\*\shell\pin" -Force | Out-Null
Set-ItemProperty -LiteralPath "HKCU:Software\Classes\*\shell\pin" -Name "ExplorerCommandHandler" -Type String -Value $pinHandler

# Pin / unpin the COM object
$cplNS.Self.InvokeVerb("pin")

# Remove the handler
Remove-Item -LiteralPath "HKCU:Software\Classes\*\shell\pin" -Recurse
```

The problem is that the item is not a *standard* directory item, but a COM class object and as such it doesn't expose the custom *verb*. Simply put - it cannot be pinned using this method.

**Approach 2** - Create a shortcut to COM object and pin the shortcut instead.  
This time, I've created a *.lnk shortcut for the COM object and tried to invoke the verb on the shortcut.
```powershell
# Get "Control Panel" namespace and localized name
$cplCLSID = "::{26EE0668-A00A-44D7-9371-BEB064C98683}"
$cplNS = (New-Object -Com "Shell.Application").NameSpace("shell:$cplCLSID")
$lnkName = $cplNS.Title

# Create a shortcut in taskbar pins folder
$lnkPath = "$env:AppData\Microsoft\Internet Explorer\Quick Launch\User Pinned\TaskBar\$lnkName.lnk"
$lnk = (New-Object -ComObject "WScript.Shell").CreateShortcut($lnkPath)
$lnk.TargetPath = $cplNS.Self.Path
$lnk.Save()

# Pin / unpin the shortcut
$lnkItem = Get-Item $lnkPath
$lnkItem = (New-Object -Com "Shell.Application").Namespace($lnkItem.DirectoryName).ParseName($lnkItem.Name)
$lnkItem.InvokeVerb("pin")

# Remove the handler
Remove-Item -LiteralPath "HKCU:Software\Classes\*\shell\pin" -Recurse
```

Unfortunately this didn't bear fruit either as there are 3 or 4 different types of shortcuts and this approach creates the one which is a direct link to the COM object. Which effectively transforms this into the first case and this pinning method can't be used.

**Approach 3** - Create a shortcut to `explorer.exe` with the COM object as path.  
In this scenario, I've tried to avoid linking the COM object directly and instead created a shortcut to `explorer.exe` and gave the CLSID as parameter.

```powershell
# Get "Control Panel" namespace and localized name
$cplCLSID = "::{26EE0668-A00A-44D7-9371-BEB064C98683}"
$cplNS = (New-Object -Com "Shell.Application").NameSpace("shell:$cplCLSID")
$lnkName = $cplNS.Title

# Create a shortcut in Quick Launch folder
$lnkPath = "$env:AppData\Microsoft\Internet Explorer\Quick Launch\$lnkName.lnk"
$lnk = (New-Object -ComObject "WScript.Shell").CreateShortcut($lnkPath)
$lnk.IconLocation = "$env:WinDir\System32\shell32.dll, 21"
$lnk.TargetPath = "$env:WinDir\explorer.exe"
$lnk.Arguments = "$cplCLSID"
$lnk.Save()

# Bring pinning / unpinning handler into '*' class scope
$pinHandler = Get-ItemPropertyValue -Path "HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\CommandStore\shell\Windows.taskbarpin" -Name "ExplorerCommandHandler"
New-Item -Path "HKCU:Software\Classes\*\shell\pin" -Force | Out-Null
Set-ItemProperty -LiteralPath "HKCU:Software\Classes\*\shell\pin" -Name "ExplorerCommandHandler" -Type String -Value $pinHandler

# Pin / unpin the shortcut
$lnkItem = Get-Item $lnkPath
$lnkItem = (New-Object -Com "Shell.Application").Namespace($lnkItem.DirectoryName).ParseName($lnkItem.Name)
$lnkItem.InvokeVerb("pin")

# Remove the handler
Remove-Item -LiteralPath "HKCU:Software\Classes\*\shell\pin" -Recurse
```
The upside is, that this approach works and allows me to pin the shortcut. However when you click the shortcut, Explorer execs into the COM object, which creates a duplicate icon on the task bar. So even if this gets where you want, it's hardly ideal as it doesn't achieve the same result as manually pinning the Control Panel window.

**Conclusion:**  
I give up.

I'm fairly certain that it *can* be doe programatically, however I don't want to spend more time on this, especially if it won't make it into the tweak library anyway. As a personal preference, I do most of the things in Windows using keyboard, so I'm opening Control Panel by typing `control` in *Run* window opened via *Win+R* keyboard shortcut. I know that this isn't what you've asked for, but it's the best I can provide.
